<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getBlast</title>
  <meta name="keywords" content="getBlast">
  <meta name="description" content="getBlast">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getBlast.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getBlast
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getBlast</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function blastStructure=getBlast(organismID,fastaFile,modelIDs,refFastaFiles) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getBlast
   Performs a bidirectional BLASTp between the organism of interest and a
   set of template organisms.

   organismID      the id of the organism of interest. This should also
                   match with the id supplied to getModelFromHomology
   fastaFile       a FASTA file with the protein sequences for the
                   organism of interest
   modelIDs        a cell array of model id:s. These must match the
                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the output
                   is to be used with getModelFromHomology
   refFastaFiles   a cell array with the paths to the corresponding FASTA
                   files
   
   blastStructure  structure containing the bidirectional homology
                   measurements which are used by getModelFromHomology

   NOTE: This function calls BLASTp to perform a bidirectional homology
   test between the organism of interest and a set of other organisms
   using standard settings. If you would like to use other homology
   measurements, please see getBlastFromExcel.

   Usage: blastStructure=getBlast(organismID,fastaFile,modelIDs,...
           refFastaFiles)

   Eduard Kerkhoven, 2017-10-20</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function blastStructure=getBlast(organismID,fastaFile,modelIDs,refFastaFiles)</a>
0002 <span class="comment">% getBlast</span>
0003 <span class="comment">%   Performs a bidirectional BLASTp between the organism of interest and a</span>
0004 <span class="comment">%   set of template organisms.</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   organismID      the id of the organism of interest. This should also</span>
0007 <span class="comment">%                   match with the id supplied to getModelFromHomology</span>
0008 <span class="comment">%   fastaFile       a FASTA file with the protein sequences for the</span>
0009 <span class="comment">%                   organism of interest</span>
0010 <span class="comment">%   modelIDs        a cell array of model id:s. These must match the</span>
0011 <span class="comment">%                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the output</span>
0012 <span class="comment">%                   is to be used with getModelFromHomology</span>
0013 <span class="comment">%   refFastaFiles   a cell array with the paths to the corresponding FASTA</span>
0014 <span class="comment">%                   files</span>
0015 <span class="comment">%</span>
0016 <span class="comment">%   blastStructure  structure containing the bidirectional homology</span>
0017 <span class="comment">%                   measurements which are used by getModelFromHomology</span>
0018 <span class="comment">%</span>
0019 <span class="comment">%   NOTE: This function calls BLASTp to perform a bidirectional homology</span>
0020 <span class="comment">%   test between the organism of interest and a set of other organisms</span>
0021 <span class="comment">%   using standard settings. If you would like to use other homology</span>
0022 <span class="comment">%   measurements, please see getBlastFromExcel.</span>
0023 <span class="comment">%</span>
0024 <span class="comment">%   Usage: blastStructure=getBlast(organismID,fastaFile,modelIDs,...</span>
0025 <span class="comment">%           refFastaFiles)</span>
0026 <span class="comment">%</span>
0027 <span class="comment">%   Eduard Kerkhoven, 2017-10-20</span>
0028 <span class="comment">%</span>
0029 
0030 <span class="comment">%Everything should be cell arrays</span>
0031 organismID=cellstr(organismID);
0032 fastaFile=cellstr(fastaFile);
0033 modelIDs=cellstr(modelIDs);
0034 refFastaFiles=cellstr(refFastaFiles);
0035 
0036 blastStructure=[];
0037 
0038 <span class="comment">%Get the directory for RAVEN Toolbox. This may not be the easiest or best</span>
0039 <span class="comment">%way to do this</span>
0040 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0041 ravenPath=fileparts(fileparts(ST(I).file));
0042 
0043 <span class="comment">%Construct databases and output file</span>
0044 tmpDB=tempname;
0045 outFile=tempname;
0046 
0047 <span class="comment">%Create a database for the new organism and blast each of the</span>
0048 <span class="comment">%refFastaFiles against it</span>
0049 
0050 <span class="keyword">if</span> isunix
0051     <span class="keyword">if</span> ismac
0052         binEnd=<span class="string">'.mac'</span>;
0053     <span class="keyword">else</span>
0054         binEnd=<span class="string">''</span>;
0055     <span class="keyword">end</span>
0056 <span class="keyword">elseif</span> ispc
0057     binEnd=<span class="string">''</span>;
0058 <span class="keyword">else</span> 
0059     dispEM(<span class="string">'Unknown OS, exiting.'</span>)
0060     <span class="keyword">return</span>
0061 <span class="keyword">end</span>
0062 
0063 <span class="comment">% Run BLAST multi-threaded to use all logical cores assigned to MATLAB.</span>
0064 cores = evalc(<span class="string">'feature(''numcores'')'</span>);
0065 cores = strsplit(cores, <span class="string">'MATLAB was assigned: '</span>);
0066 cores = regexp(cores{2},<span class="string">'^\d*'</span>,<span class="string">'match'</span>);
0067 cores = cores{1};
0068 
0069 [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast-2.6.0+'</span>,[<span class="string">'makeblastdb'</span> binEnd]) <span class="string">'&quot; -in &quot;'</span> fastaFile{1} <span class="string">'&quot; -out &quot;'</span> tmpDB <span class="string">'&quot; -dbtype &quot;prot&quot;'</span>]);
0070 
0071 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0072     fprintf([<span class="string">'BLASTing &quot;'</span> modelIDs{i} <span class="string">'&quot; against &quot;'</span> organismID{1} <span class="string">'&quot;..\n'</span>]);
0073     [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast-2.6.0+'</span>,[<span class="string">'blastp'</span> binEnd]) <span class="string">'&quot; -query &quot;'</span> refFastaFiles{i} <span class="string">'&quot; -out &quot;'</span> outFile <span class="string">'_'</span> num2str(i) <span class="string">'&quot; -db &quot;'</span> tmpDB <span class="string">'&quot; -evalue 10e-5 -outfmt &quot;10 qseqid sseqid evalue pident length bitscore ppos&quot; -num_threads '</span> cores]);
0074 <span class="keyword">end</span>
0075 delete([tmpDB <span class="string">'*'</span>]);
0076 
0077 <span class="comment">%Then create a database for each of the reference organisms and blast</span>
0078 <span class="comment">%the new organism against them</span>
0079 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0080     fprintf([<span class="string">'BLASTing &quot;'</span> organismID{1} <span class="string">'&quot; against &quot;'</span> modelIDs{i} <span class="string">'&quot;..\n'</span>]);
0081     [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast-2.6.0+'</span>,[<span class="string">'makeblastdb'</span> binEnd]) <span class="string">'&quot; -in &quot;'</span> refFastaFiles{i} <span class="string">'&quot; -out &quot;'</span> tmpDB <span class="string">'&quot; -dbtype &quot;prot&quot;'</span>]);
0082     [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast-2.6.0+'</span>,[<span class="string">'blastp'</span> binEnd]) <span class="string">'&quot; -query &quot;'</span> fastaFile{1} <span class="string">'&quot; -out &quot;'</span> outFile <span class="string">'_r'</span> num2str(i) <span class="string">'&quot; -db &quot;'</span> tmpDB <span class="string">'&quot; -evalue 10e-5 -outfmt &quot;10 qseqid sseqid evalue pident length bitscore ppos&quot; -num_threads '</span> cores]);
0083     delete([tmpDB <span class="string">'*'</span>]);
0084 <span class="keyword">end</span>
0085     
0086 <span class="comment">%Done with the BLAST, do the parsing of the text files</span>
0087 <span class="keyword">for</span> i=1:numel(refFastaFiles)*2
0088     tempStruct=[];
0089     <span class="keyword">if</span> i&lt;=numel(refFastaFiles)
0090         tempStruct.fromId=modelIDs{i};
0091         tempStruct.toId=organismID{1};
0092         A=importdata([outFile <span class="string">'_'</span> num2str(i)]);
0093     <span class="keyword">else</span>
0094         tempStruct.fromId=organismID{1};
0095         tempStruct.toId=modelIDs{i-numel(refFastaFiles)};
0096         A=importdata([outFile <span class="string">'_r'</span> num2str(i-numel(refFastaFiles))]);
0097     <span class="keyword">end</span>
0098     tempStruct.fromGenes=A.textdata(:,1);
0099     tempStruct.toGenes=A.textdata(:,2);
0100     tempStruct.evalue=A.data(:,1);
0101     tempStruct.identity=A.data(:,2);
0102     tempStruct.aligLen=A.data(:,3);
0103     tempStruct.bitscore=A.data(:,4);
0104     tempStruct.ppos=A.data(:,5);
0105     blastStructure=[blastStructure tempStruct];
0106 <span class="keyword">end</span>
0107 
0108 <span class="comment">%Remove the old tempfiles</span>
0109 delete([outFile <span class="string">'*'</span>]);
0110 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 01-Nov-2017 16:53:54 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>