<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getWoLFScores</title>
  <meta name="keywords" content="getWoLFScores">
  <meta name="description" content="getWoLFScores">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getWoLFScores.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getWoLFScores
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getWoLFScores</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function geneScoreStruct=getWoLFScores(inputFile,kingdom) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getWoLFScores
   Calls WoLF PSort to predict the sub-cellular localization of proteins.
   The output can be used as input to predictLocalization. This function
   is currently only available for Linux and requires PERL to be
   installed. If you want to use another predictor, see parseScores.
   The function normalizes the scores so that the best score for each gene
   is 1.0.

   Input:
   inputFile               a FASTA file with protein sequences
   kingdom                 the kingdom of the organism, 'animal', 'fungi'
                           or 'plant'

   Output:
   geneScoreStructure      a structure to be used in predictLocalization

   Usage: geneScoreStruct=getWoLFScores(inputFile,kingdom)

   Simonas Marcisauskas, 2019-08-17</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="parseScores.html" class="code" title="function geneScoreStructure=parseScores(inputFile,predictor)">parseScores</a>	parseScores</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function geneScoreStruct=getWoLFScores(inputFile,kingdom)</a>
0002 <span class="comment">% getWoLFScores</span>
0003 <span class="comment">%   Calls WoLF PSort to predict the sub-cellular localization of proteins.</span>
0004 <span class="comment">%   The output can be used as input to predictLocalization. This function</span>
0005 <span class="comment">%   is currently only available for Linux and requires PERL to be</span>
0006 <span class="comment">%   installed. If you want to use another predictor, see parseScores.</span>
0007 <span class="comment">%   The function normalizes the scores so that the best score for each gene</span>
0008 <span class="comment">%   is 1.0.</span>
0009 <span class="comment">%</span>
0010 <span class="comment">%   Input:</span>
0011 <span class="comment">%   inputFile               a FASTA file with protein sequences</span>
0012 <span class="comment">%   kingdom                 the kingdom of the organism, 'animal', 'fungi'</span>
0013 <span class="comment">%                           or 'plant'</span>
0014 <span class="comment">%</span>
0015 <span class="comment">%   Output:</span>
0016 <span class="comment">%   geneScoreStructure      a structure to be used in predictLocalization</span>
0017 <span class="comment">%</span>
0018 <span class="comment">%   Usage: geneScoreStruct=getWoLFScores(inputFile,kingdom)</span>
0019 <span class="comment">%</span>
0020 <span class="comment">%   Simonas Marcisauskas, 2019-08-17</span>
0021 <span class="comment">%</span>
0022 
0023 <span class="keyword">if</span> ~(exist(inputFile,<span class="string">'file'</span>)==2)
0024     error(<span class="string">'FASTA file %s cannot be found'</span>,string(inputFile));
0025 <span class="keyword">end</span>
0026 
0027 <span class="keyword">if</span> ~strcmp(kingdom,<span class="string">'animal'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'fungi'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'plant'</span>)
0028     EM=<span class="string">'Allowed kingdoms are &quot;animal&quot;, &quot;fungi&quot;, and &quot;plant&quot;'</span>;
0029     dispEM(EM);
0030 <span class="keyword">end</span>
0031 
0032 <span class="keyword">if</span> ispc==true
0033     EM=<span class="string">'This function currently runs only on Linux. Use parseScores if you want to use another predictor'</span>;
0034     dispEM(EM);
0035 <span class="keyword">end</span>
0036 
0037 <span class="comment">%Get the directory for RAVEN Toolbox. This may not be the easiest or best</span>
0038 <span class="comment">%way to do this</span>
0039 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0040 ravenPath=fileparts(fileparts(ST(I).file));
0041 
0042 <span class="comment">%Temporary output name</span>
0043 outFile=tempname;
0044 fid=fopen(outFile,<span class="string">'w'</span>);
0045 
0046 <span class="comment">%Do the prediction</span>
0047 [~, output]=unix([<span class="string">'perl &quot;'</span> ravenPath <span class="string">'/software/WoLFPSORT/bin/runWolfPsortSummary&quot; '</span> kingdom <span class="string">' &lt; '</span> inputFile]);
0048 
0049 <span class="comment">%Save output and call the general parser</span>
0050 fprintf(fid,output);
0051 fclose(fid);
0052 geneScoreStruct=<a href="parseScores.html" class="code" title="function geneScoreStructure=parseScores(inputFile,predictor)">parseScores</a>(outFile,<span class="string">'wolf'</span>);
0053 
0054 <span class="comment">%Clean up</span>
0055 delete(outFile);
0056 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 06-Nov-2019 21:22:42 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>