<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getKEGGModelForOrganism</title>
  <meta name="keywords" content="getKEGGModelForOrganism">
  <meta name="description" content="getKEGGModelForOrganism">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getKEGGModelForOrganism.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getKEGGModelForOrganism
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getKEGGModelForOrganism</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,minScoreRatioKO,maxPhylDist,nSequences,seqIdentity) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getKEGGModelForOrganism
   Reconstructs a genome-scale metabolic model based on protein homology to the
   orthologies in KEGG

   organismID          three or four letter abbreviation of the organism (as
                       used in KEGG). If not available, use a closely related
                       species. This is used for determing the
                       phylogenetic distance. Use 'eukaryotes' or
                       'prokaryotes' to get a model for the whole domain.
                       Only applicable if fastaFile is empty, i.e. no
                       homology search should be performed
   fastaFile           a FASTA file that contains the protein sequences of
                       the organism for which to reconstruct a model (opt,
                       if no FASTA file is supplied then a model is
                       reconstructed based only on the organism
                       abbreviation. This option ignores all settings except for
                       keepUndefinedStoich, keepIncomplete and keepGeneral)
   dataDir             directory for which to retrieve the input data.
                       Should contain a combination of these sub-folders:
                       -dataDir\keggdb
                           The KEGG database files used in 1a (see below).
                       -dataDir\fasta
                           The multi-FASTA files generated in 1b (see below)
                       -dataDir\aligned
                           The aligned FASTA files as generated in 2a (see
                           below)
                       -dataDir\hmms
                           The hidden Markov models as generated in 2b or
                           downloaded from BioMet Toolbox (see below)
                       The final directory in dataDir should be styled as
                       proXXX_keggYY or eukXXX_keggYY, indicating whether
                       the HMMs were trained on pro- or eukaryotic sequences,
                       using a sequence similarity threshold of XXX %,
                       fitting the KEGG version YY. E.g. euk100_kegg82.
                       (opt, see note about fastaFile. Note that in order to
                       rebuild the KEGG model from a database dump, as opposed to
                       using the version supplied with RAVEN, you would still need
                       to supply this)
   outDir              directory to save the results from the quering of
                       the hidden Markov models. The output is specific
                       for the input sequences and the settings used. It
                       is stored in this manner so that the function can
                       continue if interrupted or if it should run in
                       parallel. Be careful not to leave output files
                       from different organisms or runs with different
                       settings in the same folder. They will not be
                       overwritten (opt, default is a temporary dir where
                       all *.out files are deleted before and after doing
                       the reconstruction)
   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These
                       will be dealt with as two separate metabolites
                       (opt, default true)
   keepIncomplete      include reactions which have been labelled as
                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,
                       default true)
   keepGeneral         include reactions which have been labelled as
                       &quot;general reaction&quot;. These are reactions on the form
                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore
                       unsuited for modelling purposes. Note that not all
                       reactions have this type of annotation, and the
                       script will therefore not be able to remove all
                       such reactions (opt, default false)
   cutOff              significans score from HMMer needed to assign genes
                       to a KO (opt, default 10^-50)
   minScoreRatioG      a gene is only assigned to KOs for which the score
                       is &gt;=log(score)/log(best score) for that gene.
                       This is to prevent that a gene which clearly belongs
                       to one KO is assigned also to KOs with much lower
                       scores (opt, default 0.8 (lower is less strict))
   minScoreRatioKO     ignore genes in a KO if their score is
                       &lt;log(score)/log(best score in KO). This is to
                       &quot;prune&quot; KOs which have many genes and where some
                       are clearly a better fit
                       (opt, default 0.3 (lower is less strict))
   maxPhylDist         -1: only use sequences from the same domain
                       (Prokaryota, Eukaryota)
                       other (positive) value: only use sequences for
                       organisms where the phylogenetic distance is at
                       the most this large (as calculated in getPhylDist)
                       (opt, default Inf, which means that all sequences will
                       be used)
   nSequences          for each KO, use up to this many sequences from
                       the most closely related species. This is mainly to
                       speed up the alignment process for KOs with very
                       many genes. This subsampling is performed before
                       running CD-HIT (opt, default inf)
   seqIdentity         sequence identity threshold in CD-HIT, referred as
                       &quot;global sequence identity&quot; in CD-HIT User's Guide.
                       The only possible options are 1 (100 %), 0.9 (90 %)
                       and 0.5 (50 %). If other values are provided, CD-HIT
                       is skipped (opt, default -1, i.e. CD-HIT is skipped)

   model               the reconstructed model

   PLEASE READ THIS: The input to this function can be confusing, because
   it is intended to be run in parallel on a cluster or in multiple sessions.
   It therefore saves a lot of intermediate results to storage. This also
   serves the purpose of not having to do redundant calculations. This,
   however, comes with the disadvantage of somewhat trickier handling.
   This is what this function does:

   1a. Loads files from a local KEGG FTP dump and constructs a general
       RAVEN model representing the metabolic network. KEGG FTP access
       requires a &lt;a href=&quot;matlab: web('http://www.bioinformatics.jp/en/keggftp.html')&quot;&gt;license&lt;/a&gt;.
   1b. Generates FASTA files from the KEGG FTP dump (see 1a). One
       multi-FASTA file for each KO in KEGG is generated. Make sure you
       actually need these first, as the parsing is time-consuming (see below).

   These steps only have to be re-done every time KEGG updates their
   database (or rather when the updates are large enough to warrant
   rerunning this part). Many user would probably never use this feature.

   2a. Does alignment of the multi-FASTA files for future use. This uses the
       settings &quot;useEvDist&quot; and &quot;nSequences&quot; to control which sequences
       should be used for constructing Hidden Markov models (HMMs), and
       later for matching your sequences to.
   2b. Trains hidden Markov models using HMMer for each of the aligned
       FASTA files.

   The most common alternatives here would be to use sequences from only
   eukaryotes, only prokaryotes or all sequences in KEGG. As explained in
   the README.md file, various sets of pre-trained hidden Markov models are
   available at &lt;a href=&quot;matlab: web('http://biomet-toolbox.org/index.php?page=downtools-raven')&quot;&gt;BioMet Toolbox&lt;/a&gt;.
   This is normally the most convenient way, but if you would like to use,
   for example, only fungal sequences for training the HMMs then you need
   to run this part.

   3a. Queries the HMMs with sequences for the organism you are making a
       model for. This step uses both the output from step 1a and from 2b.
   3b. Constructs a model based on the fit to the HMMs and the chosen
       parameters.

   These steps are specific to the organism for which you are reconstructing
   the model.

   In principle the function looks at which output that is already available
   and runs only the parts that are required for step 3. This means
   that (see the definition of the parameters for details):
   -1a is only performed if there are no KEGG model files in the
   RAVEN\external\kegg directory
   -1b is only performed if not all required HMMs OR aligned FASTA files
   OR multi-FASTA files exist in the defined dataDir. This means that this
   step is skipped if the HMMs are downloaded from BioMet Toolbox instead
   (see above). If not all files exist it will try to find
   the KEGG database files in dataDir.
   -2a is only performed if not all required HMMs OR aligned FASTA files
   files exist in the defined dataDir. This means that this step is skipped
   if the HMMs are downloaded from BioMet Toolbox instead (see above).
   -2b is only performed if not all required HMMs exist in the defined
   dataDir. This means that this step is skipped if the FASTA files or
   HMMs are downloaded from BioMet Toolbox instead (see above).
   -3a is performed for the required HMMs for which no corresponding .out
   file exists in outDir. This is just a way to enable the function to be
   run in parallel or to resume if interrupted.
   -3b is always performed.

   Usage: model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,...
    keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,...
    minScoreRatioKO,maxPhylDist,nSequences,seqIdentity)

   Eduard Kerkhoven, 2017-07-24</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getModelFromKEGG.html" class="code" title="function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>	getModelFromKEGG</li><li><a href="getPhylDist.html" class="code" title="function phylDistStruct=getPhylDist(keggPath,onlyInKingdom)">getPhylDist</a>	getPhylDist</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function files=listFiles(directory)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,</a><span class="keyword">...</span>
0002     keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,<span class="keyword">...</span>
0003     minScoreRatioKO,maxPhylDist,nSequences,seqIdentity)
0004 <span class="comment">% getKEGGModelForOrganism</span>
0005 <span class="comment">%   Reconstructs a genome-scale metabolic model based on protein homology to the</span>
0006 <span class="comment">%   orthologies in KEGG</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   organismID          three or four letter abbreviation of the organism (as</span>
0009 <span class="comment">%                       used in KEGG). If not available, use a closely related</span>
0010 <span class="comment">%                       species. This is used for determing the</span>
0011 <span class="comment">%                       phylogenetic distance. Use 'eukaryotes' or</span>
0012 <span class="comment">%                       'prokaryotes' to get a model for the whole domain.</span>
0013 <span class="comment">%                       Only applicable if fastaFile is empty, i.e. no</span>
0014 <span class="comment">%                       homology search should be performed</span>
0015 <span class="comment">%   fastaFile           a FASTA file that contains the protein sequences of</span>
0016 <span class="comment">%                       the organism for which to reconstruct a model (opt,</span>
0017 <span class="comment">%                       if no FASTA file is supplied then a model is</span>
0018 <span class="comment">%                       reconstructed based only on the organism</span>
0019 <span class="comment">%                       abbreviation. This option ignores all settings except for</span>
0020 <span class="comment">%                       keepUndefinedStoich, keepIncomplete and keepGeneral)</span>
0021 <span class="comment">%   dataDir             directory for which to retrieve the input data.</span>
0022 <span class="comment">%                       Should contain a combination of these sub-folders:</span>
0023 <span class="comment">%                       -dataDir\keggdb</span>
0024 <span class="comment">%                           The KEGG database files used in 1a (see below).</span>
0025 <span class="comment">%                       -dataDir\fasta</span>
0026 <span class="comment">%                           The multi-FASTA files generated in 1b (see below)</span>
0027 <span class="comment">%                       -dataDir\aligned</span>
0028 <span class="comment">%                           The aligned FASTA files as generated in 2a (see</span>
0029 <span class="comment">%                           below)</span>
0030 <span class="comment">%                       -dataDir\hmms</span>
0031 <span class="comment">%                           The hidden Markov models as generated in 2b or</span>
0032 <span class="comment">%                           downloaded from BioMet Toolbox (see below)</span>
0033 <span class="comment">%                       The final directory in dataDir should be styled as</span>
0034 <span class="comment">%                       proXXX_keggYY or eukXXX_keggYY, indicating whether</span>
0035 <span class="comment">%                       the HMMs were trained on pro- or eukaryotic sequences,</span>
0036 <span class="comment">%                       using a sequence similarity threshold of XXX %,</span>
0037 <span class="comment">%                       fitting the KEGG version YY. E.g. euk100_kegg82.</span>
0038 <span class="comment">%                       (opt, see note about fastaFile. Note that in order to</span>
0039 <span class="comment">%                       rebuild the KEGG model from a database dump, as opposed to</span>
0040 <span class="comment">%                       using the version supplied with RAVEN, you would still need</span>
0041 <span class="comment">%                       to supply this)</span>
0042 <span class="comment">%   outDir              directory to save the results from the quering of</span>
0043 <span class="comment">%                       the hidden Markov models. The output is specific</span>
0044 <span class="comment">%                       for the input sequences and the settings used. It</span>
0045 <span class="comment">%                       is stored in this manner so that the function can</span>
0046 <span class="comment">%                       continue if interrupted or if it should run in</span>
0047 <span class="comment">%                       parallel. Be careful not to leave output files</span>
0048 <span class="comment">%                       from different organisms or runs with different</span>
0049 <span class="comment">%                       settings in the same folder. They will not be</span>
0050 <span class="comment">%                       overwritten (opt, default is a temporary dir where</span>
0051 <span class="comment">%                       all *.out files are deleted before and after doing</span>
0052 <span class="comment">%                       the reconstruction)</span>
0053 <span class="comment">%   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These</span>
0054 <span class="comment">%                       will be dealt with as two separate metabolites</span>
0055 <span class="comment">%                       (opt, default true)</span>
0056 <span class="comment">%   keepIncomplete      include reactions which have been labelled as</span>
0057 <span class="comment">%                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,</span>
0058 <span class="comment">%                       default true)</span>
0059 <span class="comment">%   keepGeneral         include reactions which have been labelled as</span>
0060 <span class="comment">%                       &quot;general reaction&quot;. These are reactions on the form</span>
0061 <span class="comment">%                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore</span>
0062 <span class="comment">%                       unsuited for modelling purposes. Note that not all</span>
0063 <span class="comment">%                       reactions have this type of annotation, and the</span>
0064 <span class="comment">%                       script will therefore not be able to remove all</span>
0065 <span class="comment">%                       such reactions (opt, default false)</span>
0066 <span class="comment">%   cutOff              significans score from HMMer needed to assign genes</span>
0067 <span class="comment">%                       to a KO (opt, default 10^-50)</span>
0068 <span class="comment">%   minScoreRatioG      a gene is only assigned to KOs for which the score</span>
0069 <span class="comment">%                       is &gt;=log(score)/log(best score) for that gene.</span>
0070 <span class="comment">%                       This is to prevent that a gene which clearly belongs</span>
0071 <span class="comment">%                       to one KO is assigned also to KOs with much lower</span>
0072 <span class="comment">%                       scores (opt, default 0.8 (lower is less strict))</span>
0073 <span class="comment">%   minScoreRatioKO     ignore genes in a KO if their score is</span>
0074 <span class="comment">%                       &lt;log(score)/log(best score in KO). This is to</span>
0075 <span class="comment">%                       &quot;prune&quot; KOs which have many genes and where some</span>
0076 <span class="comment">%                       are clearly a better fit</span>
0077 <span class="comment">%                       (opt, default 0.3 (lower is less strict))</span>
0078 <span class="comment">%   maxPhylDist         -1: only use sequences from the same domain</span>
0079 <span class="comment">%                       (Prokaryota, Eukaryota)</span>
0080 <span class="comment">%                       other (positive) value: only use sequences for</span>
0081 <span class="comment">%                       organisms where the phylogenetic distance is at</span>
0082 <span class="comment">%                       the most this large (as calculated in getPhylDist)</span>
0083 <span class="comment">%                       (opt, default Inf, which means that all sequences will</span>
0084 <span class="comment">%                       be used)</span>
0085 <span class="comment">%   nSequences          for each KO, use up to this many sequences from</span>
0086 <span class="comment">%                       the most closely related species. This is mainly to</span>
0087 <span class="comment">%                       speed up the alignment process for KOs with very</span>
0088 <span class="comment">%                       many genes. This subsampling is performed before</span>
0089 <span class="comment">%                       running CD-HIT (opt, default inf)</span>
0090 <span class="comment">%   seqIdentity         sequence identity threshold in CD-HIT, referred as</span>
0091 <span class="comment">%                       &quot;global sequence identity&quot; in CD-HIT User's Guide.</span>
0092 <span class="comment">%                       The only possible options are 1 (100 %), 0.9 (90 %)</span>
0093 <span class="comment">%                       and 0.5 (50 %). If other values are provided, CD-HIT</span>
0094 <span class="comment">%                       is skipped (opt, default -1, i.e. CD-HIT is skipped)</span>
0095 <span class="comment">%</span>
0096 <span class="comment">%   model               the reconstructed model</span>
0097 <span class="comment">%</span>
0098 <span class="comment">%   PLEASE READ THIS: The input to this function can be confusing, because</span>
0099 <span class="comment">%   it is intended to be run in parallel on a cluster or in multiple sessions.</span>
0100 <span class="comment">%   It therefore saves a lot of intermediate results to storage. This also</span>
0101 <span class="comment">%   serves the purpose of not having to do redundant calculations. This,</span>
0102 <span class="comment">%   however, comes with the disadvantage of somewhat trickier handling.</span>
0103 <span class="comment">%   This is what this function does:</span>
0104 <span class="comment">%</span>
0105 <span class="comment">%   1a. Loads files from a local KEGG FTP dump and constructs a general</span>
0106 <span class="comment">%       RAVEN model representing the metabolic network. KEGG FTP access</span>
0107 <span class="comment">%       requires a &lt;a href=&quot;matlab: web('http://www.bioinformatics.jp/en/keggftp.html')&quot;&gt;license&lt;/a&gt;.</span>
0108 <span class="comment">%   1b. Generates FASTA files from the KEGG FTP dump (see 1a). One</span>
0109 <span class="comment">%       multi-FASTA file for each KO in KEGG is generated. Make sure you</span>
0110 <span class="comment">%       actually need these first, as the parsing is time-consuming (see below).</span>
0111 <span class="comment">%</span>
0112 <span class="comment">%   These steps only have to be re-done every time KEGG updates their</span>
0113 <span class="comment">%   database (or rather when the updates are large enough to warrant</span>
0114 <span class="comment">%   rerunning this part). Many user would probably never use this feature.</span>
0115 <span class="comment">%</span>
0116 <span class="comment">%   2a. Does alignment of the multi-FASTA files for future use. This uses the</span>
0117 <span class="comment">%       settings &quot;useEvDist&quot; and &quot;nSequences&quot; to control which sequences</span>
0118 <span class="comment">%       should be used for constructing Hidden Markov models (HMMs), and</span>
0119 <span class="comment">%       later for matching your sequences to.</span>
0120 <span class="comment">%   2b. Trains hidden Markov models using HMMer for each of the aligned</span>
0121 <span class="comment">%       FASTA files.</span>
0122 <span class="comment">%</span>
0123 <span class="comment">%   The most common alternatives here would be to use sequences from only</span>
0124 <span class="comment">%   eukaryotes, only prokaryotes or all sequences in KEGG. As explained in</span>
0125 <span class="comment">%   the README.md file, various sets of pre-trained hidden Markov models are</span>
0126 <span class="comment">%   available at &lt;a href=&quot;matlab: web('http://biomet-toolbox.org/index.php?page=downtools-raven')&quot;&gt;BioMet Toolbox&lt;/a&gt;.</span>
0127 <span class="comment">%   This is normally the most convenient way, but if you would like to use,</span>
0128 <span class="comment">%   for example, only fungal sequences for training the HMMs then you need</span>
0129 <span class="comment">%   to run this part.</span>
0130 <span class="comment">%</span>
0131 <span class="comment">%   3a. Queries the HMMs with sequences for the organism you are making a</span>
0132 <span class="comment">%       model for. This step uses both the output from step 1a and from 2b.</span>
0133 <span class="comment">%   3b. Constructs a model based on the fit to the HMMs and the chosen</span>
0134 <span class="comment">%       parameters.</span>
0135 <span class="comment">%</span>
0136 <span class="comment">%   These steps are specific to the organism for which you are reconstructing</span>
0137 <span class="comment">%   the model.</span>
0138 <span class="comment">%</span>
0139 <span class="comment">%   In principle the function looks at which output that is already available</span>
0140 <span class="comment">%   and runs only the parts that are required for step 3. This means</span>
0141 <span class="comment">%   that (see the definition of the parameters for details):</span>
0142 <span class="comment">%   -1a is only performed if there are no KEGG model files in the</span>
0143 <span class="comment">%   RAVEN\external\kegg directory</span>
0144 <span class="comment">%   -1b is only performed if not all required HMMs OR aligned FASTA files</span>
0145 <span class="comment">%   OR multi-FASTA files exist in the defined dataDir. This means that this</span>
0146 <span class="comment">%   step is skipped if the HMMs are downloaded from BioMet Toolbox instead</span>
0147 <span class="comment">%   (see above). If not all files exist it will try to find</span>
0148 <span class="comment">%   the KEGG database files in dataDir.</span>
0149 <span class="comment">%   -2a is only performed if not all required HMMs OR aligned FASTA files</span>
0150 <span class="comment">%   files exist in the defined dataDir. This means that this step is skipped</span>
0151 <span class="comment">%   if the HMMs are downloaded from BioMet Toolbox instead (see above).</span>
0152 <span class="comment">%   -2b is only performed if not all required HMMs exist in the defined</span>
0153 <span class="comment">%   dataDir. This means that this step is skipped if the FASTA files or</span>
0154 <span class="comment">%   HMMs are downloaded from BioMet Toolbox instead (see above).</span>
0155 <span class="comment">%   -3a is performed for the required HMMs for which no corresponding .out</span>
0156 <span class="comment">%   file exists in outDir. This is just a way to enable the function to be</span>
0157 <span class="comment">%   run in parallel or to resume if interrupted.</span>
0158 <span class="comment">%   -3b is always performed.</span>
0159 <span class="comment">%</span>
0160 <span class="comment">%   Usage: model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,...</span>
0161 <span class="comment">%    keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,...</span>
0162 <span class="comment">%    minScoreRatioKO,maxPhylDist,nSequences,seqIdentity)</span>
0163 <span class="comment">%</span>
0164 <span class="comment">%   Eduard Kerkhoven, 2017-07-24</span>
0165 <span class="comment">%</span>
0166 
0167 <span class="keyword">if</span> nargin&lt;2
0168     fastaFile=[];
0169 <span class="keyword">end</span>
0170 <span class="keyword">if</span> nargin&lt;3
0171     dataDir=[];
0172 <span class="keyword">end</span>
0173 <span class="keyword">if</span> nargin&lt;4
0174     outDir=[];
0175 <span class="keyword">end</span>
0176 <span class="keyword">if</span> isempty(outDir)
0177     outDir=tempdir;
0178     <span class="comment">%Delete all *.out files if any exist</span>
0179     delete(fullfile(outDir,<span class="string">'*.out'</span>));
0180 <span class="keyword">end</span>
0181 <span class="keyword">if</span> nargin&lt;5
0182     keepUndefinedStoich=true;
0183 <span class="keyword">end</span>
0184 <span class="keyword">if</span> nargin&lt;6
0185     keepIncomplete=true;
0186 <span class="keyword">end</span>
0187 <span class="keyword">if</span> nargin&lt;7
0188     keepGeneral=false;
0189 <span class="keyword">end</span>
0190 <span class="keyword">if</span> nargin&lt;8
0191     cutOff=10^-50;
0192 <span class="keyword">end</span>
0193 <span class="keyword">if</span> nargin&lt;9
0194     minScoreRatioG=0.8;
0195 <span class="keyword">end</span>
0196 <span class="keyword">if</span> nargin&lt;10
0197     minScoreRatioKO=0.3;
0198 <span class="keyword">end</span>
0199 <span class="keyword">if</span> nargin&lt;11
0200     maxPhylDist=inf; <span class="comment">%Include all sequences for each reaction</span>
0201 <span class="keyword">end</span>
0202 <span class="keyword">if</span> nargin&lt;12
0203     nSequences=inf; <span class="comment">%Include all sequences for each reaction</span>
0204 <span class="keyword">end</span>
0205 <span class="keyword">if</span> nargin&lt;13
0206     seqIdentity=-1; <span class="comment">%CD-HIT is not used in the pipeline</span>
0207 <span class="keyword">end</span>
0208 
0209 <span class="comment">% Checking if dataDir is consistent. It must point to pre-trained HMMs set,</span>
0210 <span class="comment">% compatible with the the current RAVEN version.</span>
0211 <span class="comment">% The user may have the required zip file already in working directory or</span>
0212 <span class="comment">% have it extracted. If the zip file and directory is not here, it is</span>
0213 <span class="comment">% downloaded from BioMet ToolBox 2.0 server.</span>
0214 
0215 <span class="keyword">if</span> ~isempty(dataDir)
0216     hmmOptions={<span class="string">'euk100_kegg82'</span>; <span class="keyword">...</span>
0217         <span class="string">'euk90_kegg82'</span>; <span class="keyword">...</span>
0218         <span class="string">'euk50_kegg82'</span>; <span class="keyword">...</span>
0219         <span class="string">'prok100_kegg82'</span>; <span class="keyword">...</span>
0220         <span class="string">'prok90_kegg82'</span>; <span class="keyword">...</span>
0221         <span class="string">'prok50_kegg82'</span>};
0222     <span class="keyword">if</span> isempty(regexp(dataDir,strcat(hmmOptions,<span class="string">'$'</span>))) <span class="comment">% Check if dataDir ends with any of the hmmOptions</span>
0223         EM=<span class="string">'Pre-trained HMMs set is not recognised. It should match any of the following sets (which are available to download):'</span>;
0224         disp(EM);
0225         disp(hmmOptions);
0226         <span class="keyword">return</span>;
0227     <span class="keyword">end</span>;
0228     <span class="keyword">if</span> exist(dataDir,<span class="string">'dir'</span>)
0229         fprintf(<span class="string">'Provided dataDir is in correct format for this RAVEN version in order to use pre-trained HMMs...\n'</span>)
0230     <span class="keyword">elseif</span> ~exist(dataDir,<span class="string">'dir'</span>) &amp;&amp; exist([dataDir,<span class="string">'.zip'</span>],<span class="string">'file'</span>)
0231         fprintf(<span class="string">'Extracting HMMs archive file...\n'</span>);
0232         unzip([dataDir,<span class="string">'.zip'</span>]);
0233     <span class="keyword">else</span>
0234         fprintf(<span class="string">'Downloading HMMs archive file...\n'</span>);
0235         websave([dataDir,<span class="string">'.zip'</span>],[<span class="string">'http://biomet-toolbox.org/tools/downloadable/files/'</span>,dataDir,<span class="string">'.zip'</span>]);
0236         fprintf(<span class="string">'Extracting HMMs archive file...\n'</span>);
0237         unzip([dataDir,<span class="string">'.zip'</span>]);
0238     <span class="keyword">end</span>;
0239 <span class="keyword">end</span>;
0240 
0241 <span class="comment">%Check if the fasta-file contains '/' or'\'. If not then it's probably just</span>
0242 <span class="comment">%a file name. It is then merged with the current folder</span>
0243 <span class="keyword">if</span> any(fastaFile)
0244     <span class="keyword">if</span> ~any(strfind(fastaFile,<span class="string">'\'</span>)) &amp;&amp; ~any(strfind(fastaFile,<span class="string">'/'</span>))
0245         fastaFile=fullfile(pwd,fastaFile);
0246     <span class="keyword">end</span>
0247     <span class="comment">%Create the required sub-folders in dataDir if they dont exist</span>
0248     <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'keggdb'</span>),<span class="string">'dir'</span>)
0249         mkdir(dataDir,<span class="string">'keggdb'</span>);
0250     <span class="keyword">end</span>
0251     <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'fasta'</span>),<span class="string">'dir'</span>)
0252         mkdir(dataDir,<span class="string">'fasta'</span>);
0253     <span class="keyword">end</span>
0254     <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'aligned'</span>),<span class="string">'dir'</span>)
0255         mkdir(dataDir,<span class="string">'aligned'</span>);
0256     <span class="keyword">end</span>
0257     <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'hmms'</span>),<span class="string">'dir'</span>)
0258         mkdir(dataDir,<span class="string">'hmms'</span>);
0259     <span class="keyword">end</span>
0260     <span class="keyword">if</span> ~exist(outDir,<span class="string">'dir'</span>)
0261         mkdir(outDir);
0262     <span class="keyword">end</span>
0263 <span class="keyword">end</span>
0264 
0265 <span class="comment">%First generate the full KEGG model. The dataDir mustn't be supplied as</span>
0266 <span class="comment">%there is also an internal RAVEN version available</span>
0267 <span class="keyword">if</span> any(dataDir)
0268     [model, KOModel]=<a href="getModelFromKEGG.html" class="code" title="function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>(fullfile(dataDir,<span class="string">'keggdb'</span>),keepUndefinedStoich,keepIncomplete,keepGeneral);
0269 <span class="keyword">else</span>
0270     [model, KOModel]=<a href="getModelFromKEGG.html" class="code" title="function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>([],keepUndefinedStoich,keepIncomplete,keepGeneral);
0271 <span class="keyword">end</span>
0272 fprintf(<span class="string">'Completed generation of KEGG model\n'</span>);
0273 model.id=organismID;
0274 model.c=zeros(numel(model.rxns),1);
0275 
0276 <span class="comment">%If no FASTA file is supplied, then just remove all genes which are not for</span>
0277 <span class="comment">%the given organism ID</span>
0278 <span class="keyword">if</span> isempty(fastaFile)
0279     <span class="keyword">if</span> ismember(organismID,{<span class="string">'eukaryotes'</span>,<span class="string">'prokaryotes'</span>})
0280         phylDists=<a href="getPhylDist.html" class="code" title="function phylDistStruct=getPhylDist(keggPath,onlyInKingdom)">getPhylDist</a>(fullfile(dataDir,<span class="string">'keggdb'</span>),maxPhylDist==-1);
0281         <span class="keyword">if</span> strcmp(organismID,<span class="string">'eukaryotes'</span>)
0282             proxyid=<span class="string">'hsa'</span>; <span class="comment">%Use H. sapiens here</span>
0283         <span class="keyword">else</span>
0284             proxyid=<span class="string">'eco'</span>; <span class="comment">%Use E. coli here</span>
0285         <span class="keyword">end</span>
0286         [~,phylDistId]=ismember(proxyid,phylDists.ids);
0287         idsToKeep=phylDists.ids(~isinf(phylDists.distMat(phylDistId,:)));
0288         taxIDs=cellfun(@(x) x{1},cellfun(@(x) strsplit(x,<span class="string">':'</span>),model.genes,<span class="string">'UniformOutput'</span>,false),<span class="string">'UniformOutput'</span>,false);
0289         I=ismember(upper(taxIDs),upper(idsToKeep));
0290     <span class="keyword">else</span>
0291         <span class="comment">% KEGG organism IDs may have three or four letters</span>
0292         organismID=strcat(organismID,<span class="string">':'</span>);   <span class="comment">%Add colon for accurate matching</span>
0293         <span class="keyword">if</span> length(organismID)==4
0294             I=cellfun(@(x) strcmpi(x(1:4),organismID),model.genes);
0295         <span class="keyword">else</span> <span class="keyword">if</span> length(organismID)==5
0296             I=cellfun(@(x) strcmpi(x(1:5),organismID),model.genes);
0297             <span class="keyword">end</span>
0298         <span class="keyword">end</span>
0299     <span class="keyword">end</span>
0300 
0301     <span class="comment">%Remove those genes</span>
0302     model.genes=model.genes(I);
0303     model.rxnGeneMat=model.rxnGeneMat(:,I);
0304 <span class="keyword">end</span>
0305 
0306 <span class="comment">%First remove all reactions without genes</span>
0307 hasGenes=any(model.rxnGeneMat,2);
0308 
0309 model=removeReactions(model,~hasGenes,true);
0310 
0311 <span class="comment">%Clean gene names</span>
0312 <span class="keyword">for</span> i=1:numel(model.genes)
0313     <span class="comment">%First get rid of the prefix organism id</span>
0314     model.genes{i}=model.genes{i}(strfind(model.genes{i},<span class="string">':'</span>)+1:end);
0315     
0316     <span class="comment">%Find and remove the description in parentheses if any</span>
0317     s=strfind(model.genes{i},<span class="string">'('</span>);
0318     <span class="keyword">if</span> any(s)
0319         model.genes{i}=model.genes{i}(1:s-1);
0320     <span class="keyword">end</span>
0321 <span class="keyword">end</span>
0322 
0323 <span class="comment">%If no FASTA file is supplied, then we're done here</span>
0324 <span class="keyword">if</span> isempty(fastaFile)
0325     <span class="comment">%Create grRules</span>
0326     model.grRules=cell(numel(model.rxns),1);
0327     model.grRules(:)={<span class="string">''</span>};
0328     
0329     <span class="comment">%Add the gene associations as 'or'</span>
0330     <span class="keyword">for</span> i=1:numel(model.rxns)
0331         <span class="comment">%Find the involved genes</span>
0332         I=find(model.rxnGeneMat(i,:));
0333         model.grRules{i}=[<span class="string">'('</span> model.genes{I(1)}];
0334         <span class="keyword">for</span> j=2:numel(I)
0335             model.grRules{i}=[model.grRules{i} <span class="string">' or '</span> model.genes{I(j)}];
0336         <span class="keyword">end</span>
0337         model.grRules{i}=[model.grRules{i} <span class="string">')'</span>];
0338     <span class="keyword">end</span>
0339     <span class="keyword">return</span>;
0340 <span class="keyword">end</span>
0341 
0342 <span class="comment">%Trim the genes so that they only contain information that can be matched</span>
0343 <span class="comment">%to the KEGG file of protein sequences (remove all information after first</span>
0344 <span class="comment">%parenthesis)</span>
0345 
0346 <span class="comment">%NOTE: For some reason the organism abbreviation should be in lower case in</span>
0347 <span class="comment">%this database. Fix this here</span>
0348 <span class="keyword">for</span> i=1:numel(KOModel.genes)
0349     parIndex=strfind(KOModel.genes{i},<span class="string">'('</span>);
0350     <span class="keyword">if</span> any(parIndex)
0351        KOModel.genes{i}=KOModel.genes{i}(1:parIndex-1);
0352     <span class="keyword">end</span>
0353     colIndex=strfind(KOModel.genes{i},<span class="string">':'</span>);
0354     KOModel.genes{i}=[lower(KOModel.genes{i}(1:colIndex-1)) KOModel.genes{i}(colIndex:end)];
0355 <span class="keyword">end</span>
0356 
0357 <span class="comment">%Create a phylogenetic distance structure</span>
0358 phylDistStruct=<a href="getPhylDist.html" class="code" title="function phylDistStruct=getPhylDist(keggPath,onlyInKingdom)">getPhylDist</a>(fullfile(dataDir,<span class="string">'keggdb'</span>),maxPhylDist==-1);
0359 [~, phylDistId]=ismember(model.id,phylDistStruct.ids);
0360 fprintf(<span class="string">'Completed creation of phylogenetic distance matrix\n'</span>);
0361 
0362 <span class="comment">%Calculate the real maximal distance now. An abitary large number of 1000 is</span>
0363 <span class="comment">%used for the &quot;all in kingdom&quot; or &quot;all sequences&quot; options. This is a bit</span>
0364 <span class="comment">%inconvenient way to do it, but it's to make it fit with some older code</span>
0365 <span class="keyword">if</span> isinf(maxPhylDist) || maxPhylDist==-1
0366     maxPhylDist=1000;
0367 <span class="keyword">end</span>
0368 
0369 <span class="comment">%Get the KO ids for which files have been generated. Maybe not the neatest</span>
0370 <span class="comment">%way..</span>
0371 fastaFiles=<a href="#_sub1" class="code" title="subfunction files=listFiles(directory)">listFiles</a>(fullfile(dataDir,<span class="string">'fasta'</span>,<span class="string">'*.fa'</span>));
0372 alignedFiles=<a href="#_sub1" class="code" title="subfunction files=listFiles(directory)">listFiles</a>(fullfile(dataDir,<span class="string">'aligned'</span>,<span class="string">'*.fa'</span>));
0373 alignedWorking=<a href="#_sub1" class="code" title="subfunction files=listFiles(directory)">listFiles</a>(fullfile(dataDir,<span class="string">'aligned'</span>,<span class="string">'*.faw'</span>));
0374 hmmFiles=<a href="#_sub1" class="code" title="subfunction files=listFiles(directory)">listFiles</a>(fullfile(dataDir,<span class="string">'hmms'</span>,<span class="string">'*.hmm'</span>));
0375 outFiles=<a href="#_sub1" class="code" title="subfunction files=listFiles(directory)">listFiles</a>(fullfile(outDir,<span class="string">'*.out'</span>));
0376 
0377 <span class="comment">%Check if multi-FASTA files should be generated. This should only be</span>
0378 <span class="comment">%performed if there are IDs in the KOModel structure that haven't been</span>
0379 <span class="comment">%parsed yet</span>
0380 missingFASTA=setdiff(KOModel.rxns,[fastaFiles;alignedFiles;hmmFiles;outFiles]);
0381 
0382 <span class="keyword">if</span> ~isempty(missingFASTA)
0383     <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'keggdb'</span>,<span class="string">'genes.pep'</span>),<span class="string">'file'</span>)
0384         EM=fprintf([<span class="string">'The file ''genes.pep'' cannot be located at '</span> strrep(dataDir,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'/ and should be downloaded from the KEGG FTP.\n'</span>]);
0385         dispEM(EM);
0386     <span class="keyword">end</span>
0387     <span class="comment">%Only construct models for KOs which don't have files already</span>
0388     fastaModel=removeReactions(KOModel,setdiff(KOModel.rxns,missingFASTA),true,true);
0389     <span class="comment">%Permute the order of the KOs in the model so that constructMultiFasta</span>
0390     <span class="comment">%can be run on several processors at once</span>
0391     fastaModel=permuteModel(fastaModel,randperm(RandStream.create(<span class="string">'mrg32k3a'</span>,<span class="string">'Seed'</span>,cputime()),numel(fastaModel.rxns)),<span class="string">'rxns'</span>);
0392     constructMultiFasta(fastaModel,fullfile(dataDir,<span class="string">'keggdb'</span>,<span class="string">'genes.pep'</span>),fullfile(dataDir,<span class="string">'fasta'</span>));
0393 <span class="keyword">end</span>
0394 fprintf(<span class="string">'Completed generation of multi-FASTA files\n'</span>);
0395 
0396 <span class="comment">%Get the directory for RAVEN Toolbox. This is to get the path to the third</span>
0397 <span class="comment">%party software used</span>
0398 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0399 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0400 
0401 <span class="keyword">if</span> isunix
0402     <span class="keyword">if</span> ismac
0403         binEnd=<span class="string">'.mac'</span>;
0404     <span class="keyword">else</span>
0405         binEnd=<span class="string">''</span>;
0406     <span class="keyword">end</span>
0407 <span class="keyword">elseif</span> ispc
0408     binEnd=<span class="string">''</span>;
0409 <span class="keyword">else</span>
0410     EM=<span class="string">'Unknown OS, exiting.'</span>;
0411     disp(EM);
0412     <span class="keyword">return</span>
0413 <span class="keyword">end</span>
0414 
0415 
0416 <span class="comment">%Check if alignment of FASTA files should be performed</span>
0417 missingAligned=setdiff(KOModel.rxns,[alignedFiles;hmmFiles;alignedWorking;outFiles]);
0418 <span class="keyword">if</span> ~isempty(missingAligned)
0419     missingAligned=missingAligned(randperm(RandStream.create(<span class="string">'mrg32k3a'</span>,<span class="string">'Seed'</span>,cputime()),numel(missingAligned)));
0420     <span class="comment">%Align all sequences using MAFFT</span>
0421     <span class="keyword">for</span> i=1:numel(missingAligned)
0422         <span class="comment">%This is checked here because it could be that it is created by a</span>
0423         <span class="comment">%parallell process. The faw-files are saved as temporary files to</span>
0424         <span class="comment">%keppt track of which files are being worked on</span>
0425         <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]),<span class="string">'file'</span>) &amp;&amp; ~exist(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.fa'</span>]),<span class="string">'file'</span>)
0426             <span class="comment">%Check that the multi-FASTA file exists. It should do so since</span>
0427             <span class="comment">%we are saving empty files as well. Print a warning and</span>
0428             <span class="comment">%continue if not.</span>
0429             <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'fasta'</span>,[missingAligned{i} <span class="string">'.fa'</span>]),<span class="string">'file'</span>)
0430                 EM=[<span class="string">'WARNING: The multi-FASTA file for '</span> missingAligned{i} <span class="string">' does not exist'</span>];
0431                 dispEM(EM,false);
0432                 <span class="keyword">continue</span>;
0433             <span class="keyword">end</span>
0434 
0435             <span class="comment">%If the multi-FASTA file is empty then save an empty aligned</span>
0436             <span class="comment">%file and continue</span>
0437             s=dir(fullfile(dataDir,<span class="string">'fasta'</span>,[missingAligned{i} <span class="string">'.fa'</span>]));
0438             <span class="keyword">if</span> s.bytes&lt;=0
0439                 fid=fopen(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.fa'</span>]),<span class="string">'w'</span>);
0440                 fclose(fid);
0441                 <span class="keyword">continue</span>;
0442             <span class="keyword">end</span>
0443 
0444             <span class="comment">%Create an empty file to prevent other threads to start to work on the same alignment</span>
0445             fid=fopen(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]),<span class="string">'w'</span>);
0446             fclose(fid);
0447 
0448             <span class="comment">%First load the FASTA file, then select up to nSequences sequences</span>
0449             <span class="comment">%of the most closely related species, apply any constraints from</span>
0450             <span class="comment">%maxPhylDist, and save it as a temporary file, and create the</span>
0451             <span class="comment">%model from that</span>
0452 
0453             fastaStruct=fastaread(fullfile(dataDir,<span class="string">'fasta'</span>,[missingAligned{i} <span class="string">'.fa'</span>]));
0454             phylDist=inf(numel(fastaStruct),1);
0455             <span class="keyword">for</span> j=1:numel(fastaStruct)
0456                <span class="comment">%Get the organism abbreviation</span>
0457                index=strfind(fastaStruct(j).Header,<span class="string">':'</span>);
0458                <span class="keyword">if</span> any(index)
0459                     abbrev=fastaStruct(j).Header(1:index(1)-1);
0460                     [~, index]=ismember(abbrev,phylDistStruct.ids);
0461                     <span class="keyword">if</span> any(index)
0462                         phylDist(j)=phylDistStruct.distMat(index(1),phylDistId);
0463                     <span class="keyword">end</span>
0464                <span class="keyword">end</span>
0465             <span class="keyword">end</span>
0466 
0467             <span class="comment">%Inf means that it should not be included</span>
0468             phylDist(phylDist&gt;maxPhylDist)=[];
0469 
0470             <span class="comment">%Sort based on phylDist</span>
0471             [~, order]=sort(phylDist);
0472 
0473             <span class="comment">%Save the first nSequences hits to a temporary FASTA file</span>
0474             <span class="keyword">if</span> nSequences&lt;=numel(fastaStruct)
0475                 fastaStruct=fastaStruct(order(1:nSequences));
0476             <span class="keyword">else</span>
0477                 fastaStruct=fastaStruct(order);
0478             <span class="keyword">end</span>
0479 
0480             <span class="comment">%Do the alignment if there are more than one sequences,</span>
0481             <span class="comment">%otherwise just save the sequence (or an empty file)</span>
0482             <span class="keyword">if</span> numel(fastaStruct)&gt;1
0483                 <span class="keyword">if</span> (seqIdentity==1 || seqIdentity==0.9 || seqIdentity==0.5)
0484                     <span class="keyword">if</span> ~ispc
0485                         <span class="keyword">if</span> seqIdentity==1
0486                             cdhitInp100=tempname;
0487                             fastawrite(cdhitInp100,fastaStruct);
0488                             tmpFile=tempname;
0489                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp100 <span class="string">'&quot; -o &quot;'</span> tmpFile <span class="string">'&quot; -c 1.0 -s 0.8 -n 5 -M 2000 -T 8'</span>]);
0490                             <span class="keyword">if</span> status~=0
0491                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0492                                 dispEM(EM);
0493                             <span class="keyword">end</span>
0494                             <span class="comment">%Remove the old tempfile</span>
0495                             <span class="keyword">if</span> exist(cdhitInp100, <span class="string">'file'</span>)
0496                                 delete([cdhitInp100 <span class="string">'*'</span>]);
0497                             <span class="keyword">end</span>
0498                         <span class="keyword">elseif</span> seqIdentity==0.9
0499                             cdhitInp100=tempname;
0500                             fastawrite(cdhitInp100,fastaStruct);
0501                             cdhitInp90=tempname;
0502                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp100 <span class="string">'&quot; -o &quot;'</span> cdhitInp90 <span class="string">'&quot; -c 1.0 -s 0.8 -n 5 -M 2000'</span>]);
0503                             <span class="keyword">if</span> status~=0
0504                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0505                                 dispEM(EM);
0506                             <span class="keyword">end</span>
0507                             <span class="comment">%Remove the old tempfile</span>
0508                             <span class="keyword">if</span> exist(cdhitInp100, <span class="string">'file'</span>)
0509                                 delete([cdhitInp100 <span class="string">'*'</span>]);
0510                             <span class="keyword">end</span>
0511                             tmpFile=tempname;
0512                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp90 <span class="string">'&quot; -o &quot;'</span> tmpFile <span class="string">'&quot; -c 0.9 -s 0.8 -n 5 -M 2000'</span>]);
0513                             <span class="keyword">if</span> status~=0
0514                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0515                                 dispEM(EM);
0516                             <span class="keyword">end</span>
0517                             <span class="comment">%Remove the old tempfile</span>
0518                             <span class="keyword">if</span> exist(cdhitInp90, <span class="string">'file'</span>)
0519                                 delete([cdhitInp90 <span class="string">'*'</span>]);
0520                             <span class="keyword">end</span>
0521                         <span class="keyword">elseif</span> seqIdentity==0.5
0522                             cdhitInp100=tempname;
0523                             fastawrite(cdhitInp100,fastaStruct);
0524                             cdhitInp90=tempname;
0525                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp100 <span class="string">'&quot; -o &quot;'</span> cdhitInp90 <span class="string">'&quot; -c 1.0 -s 0.8 -n 5 -M 2000'</span>]);
0526                             <span class="keyword">if</span> status~=0
0527                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0528                                 dispEM(EM);
0529                             <span class="keyword">end</span>
0530                             <span class="comment">%Remove the old tempfile</span>
0531                             <span class="keyword">if</span> exist(cdhitInp100, <span class="string">'file'</span>)
0532                                 delete([cdhitInp100 <span class="string">'*'</span>]);
0533                             <span class="keyword">end</span>
0534                             cdhitInp50=tempname;
0535                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp90 <span class="string">'&quot; -o &quot;'</span> cdhitInp50 <span class="string">'&quot; -c 0.9 -s 0.8 -n 5 -M 2000'</span>]);
0536                             <span class="keyword">if</span> status~=0
0537                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0538                                 dispEM(EM);
0539                             <span class="keyword">end</span>
0540                             <span class="comment">%Remove the old tempfile</span>
0541                             <span class="keyword">if</span> exist(cdhitInp90, <span class="string">'file'</span>)
0542                                 delete([cdhitInp90 <span class="string">'*'</span>]);
0543                             <span class="keyword">end</span>
0544                             tmpFile=tempname;
0545                             [status, output]=unix([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'cd-hit-v4.6.6'</span>,[<span class="string">'cd-hit'</span> binEnd]) <span class="string">'&quot; -i &quot;'</span> cdhitInp50 <span class="string">'&quot; -o &quot;'</span> tmpFile <span class="string">'&quot; -c 0.5 -s 0.8 -n 3 -M 2000'</span>]);
0546                             <span class="keyword">if</span> status~=0
0547                                 EM=[<span class="string">'Error when performing clustering of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0548                                 dispEM(EM);
0549                             <span class="keyword">end</span>
0550                             <span class="comment">%Remove the old tempfile</span>
0551                             <span class="keyword">if</span> exist(cdhitInp50, <span class="string">'file'</span>)
0552                                 delete([cdhitInp50 <span class="string">'*'</span>]);
0553                             <span class="keyword">end</span>
0554                         <span class="keyword">end</span>
0555                     <span class="keyword">else</span>
0556                         <span class="comment">% CD-HIT is not available in Windows so we just</span>
0557                         <span class="comment">% export multifasta file into temporary file;</span>
0558                         tmpFile=tempname;
0559                         fastawrite(tmpFile,fastaStruct);
0560                     <span class="keyword">end</span>
0561                 <span class="keyword">else</span>
0562                     tmpFile=tempname;
0563                     fastawrite(tmpFile,fastaStruct);
0564                 <span class="keyword">end</span>
0565                 <span class="comment">%Do the alignment for this file</span>
0566                 <span class="keyword">if</span> ~ispc
0567                     [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'mafft-7.305'</span>,[<span class="string">'mafft'</span> binEnd]) <span class="string">'&quot; --auto &quot;'</span> tmpFile <span class="string">'&quot; &gt; &quot;'</span> fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]) <span class="string">'&quot;'</span>]);
0568                 <span class="keyword">else</span>
0569                     [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'mafft-7.305'</span>,<span class="string">'mafft.bat'</span>) <span class="string">'&quot; --auto &quot;'</span> tmpFile <span class="string">'&quot; &gt; &quot;'</span> fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]) <span class="string">'&quot;'</span>]);
0570                 <span class="keyword">end</span>
0571                 <span class="keyword">if</span> status~=0
0572                     EM=[<span class="string">'Error when performing alignment of '</span> missingAligned{i} <span class="string">':\n'</span> output];
0573                     dispEM(EM);
0574                 <span class="keyword">end</span>
0575                 <span class="comment">%Remove the old tempfile</span>
0576                 <span class="keyword">if</span> exist(tmpFile, <span class="string">'file'</span>)
0577                    delete([tmpFile <span class="string">'*'</span>]);
0578                 <span class="keyword">end</span>
0579             <span class="keyword">else</span>
0580                 <span class="comment">%If there is only one sequence then it's not possible to do a</span>
0581                 <span class="comment">%multiple alignment. Just print the sequence instead. An</span>
0582                 <span class="comment">%empty file was written previously so that doesn't have to</span>
0583                 <span class="comment">%be dealt with</span>
0584                 <span class="keyword">if</span> numel(fastaStruct)==1
0585                     fastawrite(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]),fastaStruct);
0586                 <span class="keyword">end</span>
0587             <span class="keyword">end</span>
0588             <span class="comment">%Move the temporary file to the real one</span>
0589             movefile(fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.faw'</span>]),fullfile(dataDir,<span class="string">'aligned'</span>,[missingAligned{i} <span class="string">'.fa'</span>]),<span class="string">'f'</span>);
0590         <span class="keyword">end</span>
0591     <span class="keyword">end</span>
0592 <span class="keyword">end</span>
0593 
0594 <span class="keyword">if</span> ~ispc
0595     <span class="keyword">if</span> (seqIdentity==1 || seqIdentity==0.9 || seqIdentity==0.5)
0596         fprintf(<span class="string">'Completed clustering and multiple alignment of sequences\n'</span>);
0597     <span class="keyword">else</span>
0598         fprintf(<span class="string">'Completed multiple alignment of sequences. Protein clustering was not requested or incorrect seqIdentity value was used\n'</span>);
0599     <span class="keyword">end</span>
0600 <span class="keyword">else</span>
0601     <span class="keyword">if</span> (seqIdentity==1 || seqIdentity==0.9 || seqIdentity==0.5)
0602         fprintf(<span class="string">'Protein clustering was skipped, since CD-HIT is not compatible with Windows'</span>);
0603     <span class="keyword">end</span>
0604     fprintf(<span class="string">'Completed multiple alignment of sequences\n'</span>);
0605 <span class="keyword">end</span>
0606 
0607 <span class="comment">%Check if training of Hidden Markov models should be performed</span>
0608 missingHMMs=setdiff(KOModel.rxns,[hmmFiles;outFiles]);
0609 <span class="keyword">if</span> ~isempty(missingHMMs)
0610     missingHMMs=missingHMMs(randperm(RandStream.create(<span class="string">'mrg32k3a'</span>,<span class="string">'Seed'</span>,cputime()),numel(missingHMMs)));
0611 
0612     <span class="comment">%Train models for all missing KOs</span>
0613     <span class="keyword">for</span> i=1:numel(missingHMMs)
0614         <span class="comment">%This is checked here because it could be that it is created by a</span>
0615         <span class="comment">%parallell process</span>
0616         <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmm'</span>]),<span class="string">'file'</span>) &amp;&amp; ~exist(fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmw'</span>]),<span class="string">'file'</span>)
0617             <span class="comment">%Check that the aligned FASTA file exists. It could be that it</span>
0618             <span class="comment">%is still being worked on by some other instance of the</span>
0619             <span class="comment">%program (the .faw file should then exist). This should not</span>
0620             <span class="comment">%happen on a single computer. I don't throw an error, because</span>
0621             <span class="comment">%it should finalize the ones it can.</span>
0622             <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'aligned'</span>,[missingHMMs{i} <span class="string">'.fa'</span>]),<span class="string">'file'</span>)
0623                 EM=[<span class="string">'The aligned FASTA file for '</span> missingHMMs{i} <span class="string">' does not exist'</span>];
0624                 dispEM(EM,false);
0625                 <span class="keyword">continue</span>;
0626             <span class="keyword">end</span>
0627 
0628             <span class="comment">%If the multi-FASTA file is empty then save an empty aligned</span>
0629             <span class="comment">%file and continue</span>
0630             s=dir(fullfile(dataDir,<span class="string">'aligned'</span>,[missingHMMs{i} <span class="string">'.fa'</span>]));
0631             <span class="keyword">if</span> s.bytes&lt;=0
0632                 fid=fopen(fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmm'</span>]),<span class="string">'w'</span>);
0633                 fclose(fid);
0634                 <span class="keyword">continue</span>;
0635             <span class="keyword">end</span>
0636             <span class="comment">%Create a temporary file to indicate that it is working on the</span>
0637             <span class="comment">%KO. This is because hmmbuild cannot overwrite existing files.</span>
0638             fid=fopen(fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmw'</span>]),<span class="string">'w'</span>);
0639             fclose(fid);
0640 
0641             <span class="comment">%Create HMM</span>
0642             [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'hmmer-3.1b2'</span>,[<span class="string">'hmmbuild'</span> binEnd]) <span class="string">'&quot; &quot;'</span> fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmm'</span>]) <span class="string">'&quot; &quot;'</span> fullfile(dataDir,<span class="string">'aligned'</span>,[missingHMMs{i} <span class="string">'.fa'</span>]) <span class="string">'&quot;'</span>]);
0643             <span class="keyword">if</span> status~=0
0644                 EM=[<span class="string">'Error when training HMM for '</span> missingHMMs{i} <span class="string">':\n'</span> output];
0645                 dispEM(EM);
0646             <span class="keyword">end</span>
0647 
0648             <span class="comment">%Delete the temporary file</span>
0649             delete(fullfile(dataDir,<span class="string">'hmms'</span>,[missingHMMs{i} <span class="string">'.hmw'</span>]));
0650         <span class="keyword">end</span>
0651     <span class="keyword">end</span>
0652 <span class="keyword">end</span>
0653 fprintf(<span class="string">'Completed generation of HMMs\n'</span>);
0654 
0655 <span class="comment">%Check which new .out files that should be generated</span>
0656 <span class="comment">%Check if training of Hidden Markov models should be performed</span>
0657 missingOUT=setdiff(KOModel.rxns,outFiles);
0658 <span class="keyword">if</span> ~isempty(missingOUT)
0659     missingOUT=missingOUT(randperm(RandStream.create(<span class="string">'mrg32k3a'</span>,<span class="string">'Seed'</span>,cputime()),numel(missingOUT)));
0660     <span class="keyword">for</span> i=1:numel(missingOUT)
0661         <span class="comment">%This is checked here because it could be that it is created by a</span>
0662         <span class="comment">%parallell process</span>
0663         <span class="keyword">if</span> ~exist(fullfile(outDir,[missingOUT{i} <span class="string">'.out'</span>]),<span class="string">'file'</span>)
0664             <span class="comment">%Check that the HMM file exists. It should do so since</span>
0665             <span class="comment">%we are saving empty files as well. Print a warning and</span>
0666             <span class="comment">%continue if not.</span>
0667             <span class="keyword">if</span> ~exist(fullfile(dataDir,<span class="string">'hmms'</span>,[missingOUT{i} <span class="string">'.hmm'</span>]),<span class="string">'file'</span>)
0668                 EM=[<span class="string">'The HMM file for '</span> missingOUT{i} <span class="string">' does not exist'</span>];
0669                 dispEM(EM,false);
0670                 <span class="keyword">continue</span>;
0671             <span class="keyword">end</span>
0672 
0673             <span class="comment">%Save an empty file to prevent several threads working on the</span>
0674             <span class="comment">%same file</span>
0675             fid=fopen(fullfile(outDir,[missingOUT{i} <span class="string">'.out'</span>]),<span class="string">'w'</span>);
0676             fclose(fid);
0677 
0678             <span class="comment">%If the HMM file is empty then save an out file and continue</span>
0679             s=dir(fullfile(dataDir,<span class="string">'hmms'</span>,[missingOUT{i} <span class="string">'.hmm'</span>]));
0680             <span class="keyword">if</span> s.bytes&lt;=0
0681                 <span class="keyword">continue</span>;
0682             <span class="keyword">end</span>
0683 
0684             <span class="comment">%Check each gene in the input file against this model</span>
0685             [status, output]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'hmmer-3.1b2'</span>,[<span class="string">'hmmsearch'</span> binEnd]) <span class="string">'&quot; &quot;'</span> fullfile(dataDir,<span class="string">'hmms'</span>,[missingOUT{i} <span class="string">'.hmm'</span>]) <span class="string">'&quot; &quot;'</span> fastaFile <span class="string">'&quot;'</span>]);
0686             <span class="keyword">if</span> status~=0
0687                 EM=[<span class="string">'Error when querying HMM for '</span> missingOUT{i} <span class="string">':\n'</span> output];
0688                 dispEM(EM);
0689             <span class="keyword">end</span>
0690 
0691             <span class="comment">%Save the output to a file</span>
0692             fid=fopen(fullfile(outDir,[missingOUT{i} <span class="string">'.out'</span>]),<span class="string">'w'</span>);
0693             fwrite(fid,output);
0694             fclose(fid);
0695         <span class="keyword">end</span>
0696     <span class="keyword">end</span>
0697 <span class="keyword">end</span>
0698 fprintf(<span class="string">'Completed matching to HMMs\n'</span>);
0699 
0700 <span class="comment">%***Begin retrieving the output and putting together the resulting model</span>
0701 
0702 <span class="comment">%Retrieve matched genes from the HMMs</span>
0703 koGeneMat=zeros(numel(KOModel.rxns),3000); <span class="comment">%Make room for 3000 genes</span>
0704 genes=cell(3000,1);
0705 <span class="comment">%Store the best score for a gene in a hash list (since I will be searching</span>
0706 <span class="comment">%many times)</span>
0707 hTable = java.util.Hashtable;
0708 
0709 geneCounter=0;
0710 <span class="keyword">for</span> i=1:numel(KOModel.rxns)
0711     <span class="keyword">if</span> exist(fullfile(outDir,[KOModel.rxns{i} <span class="string">'.out'</span>]), <span class="string">'file'</span>)
0712         fid=fopen(fullfile(outDir,[KOModel.rxns{i} <span class="string">'.out'</span>]),<span class="string">'r'</span>);
0713         beginMatches=false;
0714         <span class="keyword">while</span> 1
0715             <span class="comment">%Get the next line</span>
0716             tline = fgetl(fid);
0717 
0718             <span class="comment">%Abort at end of file</span>
0719             <span class="keyword">if</span> ~ischar(tline)
0720                 <span class="keyword">break</span>;
0721             <span class="keyword">end</span>
0722 
0723             <span class="keyword">if</span> and(beginMatches,strcmp(tline,<span class="string">'  ------ inclusion threshold ------'</span>))
0724                 <span class="keyword">break</span>;
0725             <span class="keyword">end</span>
0726 
0727             <span class="keyword">if</span> beginMatches==false
0728                 <span class="comment">%This is how the listing of matches begins</span>
0729                 <span class="keyword">if</span> any(strfind(tline,<span class="string">'E-value '</span>))
0730                     <span class="comment">%Read one more line that is only padding</span>
0731                     tline = fgetl(fid);
0732                     beginMatches=true;
0733                 <span class="keyword">end</span>
0734             <span class="keyword">else</span>
0735                 <span class="comment">%If matches should be read</span>
0736                 <span class="keyword">if</span> ~strcmp(tline,<span class="string">'   [No hits detected that satisfy reporting thresholds]'</span>) &amp;&amp; ~isempty(tline)
0737                     elements=regexp(tline,<span class="string">' '</span>,<span class="string">'split'</span>);
0738                     elements=elements(cellfun(@any,elements));
0739 
0740                     <span class="comment">%Check if the match is below the treshhold</span>
0741                     score=str2num(elements{1});
0742                     gene=elements{9};
0743                     <span class="keyword">if</span> score&lt;=cutOff
0744                         <span class="comment">%If the score is exactly 0, change it to a very</span>
0745                         <span class="comment">%small value to avoid NaN</span>
0746                         <span class="keyword">if</span> score==0
0747                            score=10^-250;
0748                         <span class="keyword">end</span>
0749                         <span class="comment">%Check if the gene is added already and, is so, get</span>
0750                         <span class="comment">%the best score for it</span>
0751                         I=hTable.get(gene);
0752                         <span class="keyword">if</span> any(I)
0753                             koGeneMat(i,I)=score;
0754                         <span class="keyword">else</span>
0755                             geneCounter=geneCounter+1;
0756                             <span class="comment">%The gene was not present yet so add it</span>
0757                             hTable.put(gene,geneCounter);
0758                             genes{geneCounter}=gene;
0759                             koGeneMat(i,geneCounter)=score;
0760                         <span class="keyword">end</span>
0761                     <span class="keyword">end</span>
0762                 <span class="keyword">else</span>
0763                     <span class="keyword">break</span>;
0764                 <span class="keyword">end</span>
0765             <span class="keyword">end</span>
0766         <span class="keyword">end</span>
0767         fclose(fid);
0768     <span class="keyword">end</span>
0769 <span class="keyword">end</span>
0770 koGeneMat=koGeneMat(:,1:geneCounter);
0771 
0772 <span class="comment">%Remove the genes for each KO that are below minScoreRatioKO.</span>
0773 <span class="keyword">for</span> i=1:size(koGeneMat,1)
0774     J=find(koGeneMat(i,:));
0775     <span class="keyword">if</span> any(J)
0776         koGeneMat(i,J(log(koGeneMat(i,J))/log(min(koGeneMat(i,J)))&lt;minScoreRatioKO))=0;
0777     <span class="keyword">end</span>
0778 <span class="keyword">end</span>
0779 
0780 <span class="comment">%Remove the KOs for each gene that are below minScoreRatioG</span>
0781 <span class="keyword">for</span> i=1:size(koGeneMat,2)
0782     J=find(koGeneMat(:,i));
0783     <span class="keyword">if</span> any(J)
0784         koGeneMat(J(log(koGeneMat(J,i))/log(min(koGeneMat(J,i)))&lt;minScoreRatioG),i)=0;
0785     <span class="keyword">end</span>
0786 <span class="keyword">end</span>
0787 
0788 <span class="comment">%Create the new model</span>
0789 model.genes=genes(1:geneCounter);
0790 model.grRules=cell(numel(model.rxns),1);
0791 model.grRules(:)={<span class="string">''</span>};
0792 model.rxnGeneMat=sparse(numel(model.rxns),numel(model.genes));
0793 
0794 <span class="comment">%Loop through the reactions and add the corresponding genes</span>
0795 <span class="keyword">for</span> i=1:numel(model.rxns)
0796     <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0797         <span class="comment">%Get all KOs</span>
0798         I=find(strcmpi(model.rxnMiriams{i}.name,<span class="string">'kegg.orthology'</span>));
0799         KOs=model.rxnMiriams{i}.value(I);
0800         <span class="comment">%Find the KOs and the corresponding genes</span>
0801         J=ismember(KOModel.rxns,KOs);
0802         [~, K]=find(koGeneMat(J,:));
0803 
0804         <span class="keyword">if</span> any(K)
0805             model.rxnGeneMat(i,K)=1;
0806             <span class="comment">%Also delete KOs for which no genes were found. If no genes at all</span>
0807             <span class="comment">%were matched to the reaction it will be deleted later</span>
0808             L=sum(koGeneMat(J,:),2)==0;
0809             model.rxnMiriams{i}.value(I(L))=[];
0810             model.rxnMiriams{i}.name(I(L))=[];
0811         <span class="keyword">end</span>
0812     <span class="keyword">end</span>
0813 <span class="keyword">end</span>
0814 
0815 <span class="comment">%Find and delete all reactions without any genes. This also removes genes</span>
0816 <span class="comment">%that are not used (which could happen because minScoreRatioG and</span>
0817 <span class="comment">%minScoreRatioKO)</span>
0818 I=sum(model.rxnGeneMat,2)==0;
0819 model=removeReactions(model,I,true,true);
0820 
0821 <span class="comment">%Add the gene associations as 'or'</span>
0822 <span class="keyword">for</span> i=1:numel(model.rxns)
0823     <span class="comment">%Find the involved genes</span>
0824     I=find(model.rxnGeneMat(i,:));
0825     model.grRules{i}=[<span class="string">'('</span> model.genes{I(1)}];
0826     <span class="keyword">for</span> j=2:numel(I)
0827         model.grRules{i}=[model.grRules{i} <span class="string">' or '</span> model.genes{I(j)}];
0828     <span class="keyword">end</span>
0829     model.grRules{i}=[model.grRules{i} <span class="string">')'</span>];
0830 <span class="keyword">end</span>
0831 <span class="keyword">end</span>
0832 
0833 
0834 <span class="comment">%Supporter function to list the files in a directory and return them as a</span>
0835 <span class="comment">%cell array</span>
0836 <a name="_sub1" href="#_subfunctions" class="code">function files=listFiles(directory)</a>
0837     temp=dir(directory);
0838     files=cell(numel(temp),1);
0839     <span class="keyword">for</span> i=1:numel(temp)
0840         files{i}=temp(i,1).name;
0841     <span class="keyword">end</span>
0842     files=strrep(files,<span class="string">'.fa'</span>,<span class="string">''</span>);
0843     files=strrep(files,<span class="string">'.hmm'</span>,<span class="string">''</span>);
0844     files=strrep(files,<span class="string">'.out'</span>,<span class="string">''</span>);
0845     files=strrep(files,<span class="string">'.faw'</span>,<span class="string">''</span>);
0846 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 01-Nov-2017 16:53:54 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>