<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of tutorial4</title>
  <meta name="keywords" content="tutorial4">
  <meta name="description" content="This exercise is about creating a model from KEGG, based on protein sequences">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">tutorial</a> &gt; tutorial4.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for tutorial&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>tutorial4
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>This exercise is about creating a model from KEGG, based on protein sequences</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment">This exercise is about creating a model from KEGG, based on protein sequences
in a FASTA file, and doing some functionality checks on the model. The example
case is for the yeast Saccharomyces cerevisiae. This tutorial is more of a
showcase than the other three, and it's main purpose is to serve as a
scaffold if you would like to reconstruct a GEM for your own organism.</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="../core/addExchangeRxns.html" class="code" title="function [model addedRxns]=addExchangeRxns(model,reactionType,mets)">addExchangeRxns</a>	addExchangeRxns</li><li><a href="../core/canProduce.html" class="code" title="function produced=canProduce(model,mets)">canProduce</a>	canProduce</li><li><a href="../core/consumeSomething.html" class="code" title="function [solution metabolite]=consumeSomething(model,ignoreMets,isNames,minNrFluxes,params,ignoreIntBounds)">consumeSomething</a>	consumeSomething</li><li><a href="../core/fillGaps.html" class="code" title="function [newConnected cannotConnect addedRxns newModel exitFlag]=fillGaps(model,models,allowNetProduction,useModelConstraints,supressWarnings,rxnScores,params)">fillGaps</a>	fillGaps</li><li><a href="../core/gapReport.html" class="code" title="function [noFluxRxns, noFluxRxnsRelaxed, subGraphs, notProducedMets, minToConnect,neededForProductionMat, canProduceWithoutInput, canConsumeWithoutOutput,connectedFromTemplates, addedFromTemplates]=gapReport(model, templateModels)">gapReport</a>	gapReport</li><li><a href="../core/getElementalBalance.html" class="code" title="function balanceStructure=getElementalBalance(model,rxns,printUnbalanced,printUnparsable)">getElementalBalance</a>	getElementalBalance</li><li><a href="../core/makeSomething.html" class="code" title="function [solution metabolite]=makeSomething(model,ignoreMets,isNames,minNrFluxes,allowExcretion,params,ignoreIntBounds)">makeSomething</a>	makeSomething</li><li><a href="../core/printFluxes.html" class="code" title="function printFluxes(model, fluxes, onlyExchange, cutOffFlux, outputFile,outputString,metaboliteList)">printFluxes</a>	printFluxes</li><li><a href="../core/removeBadRxns.html" class="code" title="function [newModel removedRxns]=removeBadRxns(model,rxnRules,ignoreMets,isNames,balanceElements,refModel,ignoreIntBounds,printReport)">removeBadRxns</a>	removeBadRxns</li><li><a href="../external/kegg/getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,minScoreRatioKO,maxPhylDist,nSequences)">getKEGGModelForOrganism</a>	getKEGGModelForOrganism</li><li><a href="../external/kegg/getModelFromKEGG.html" class="code" title="function [model KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>	getModelFromKEGG</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">%This exercise is about creating a model from KEGG, based on protein sequences</span>
0002 <span class="comment">%in a FASTA file, and doing some functionality checks on the model. The example</span>
0003 <span class="comment">%case is for the yeast Saccharomyces cerevisiae. This tutorial is more of a</span>
0004 <span class="comment">%showcase than the other three, and it's main purpose is to serve as a</span>
0005 <span class="comment">%scaffold if you would like to reconstruct a GEM for your own organism.</span>
0006 
0007 <span class="comment">%Start by downloading trained Hidden Markov Models for eukaryots from the</span>
0008 <span class="comment">%RAVEN toolbox homepage. Unzip the content to a folder. Read the help for</span>
0009 <span class="comment">%getKEGGModelForOrganism for details regarding the folders and what they</span>
0010 <span class="comment">%are for.</span>
0011 
0012 <span class="comment">%This creates a model for S. cerevisiae. The parameters are set to exclude</span>
0013 <span class="comment">%general or unclear reactions and reactions with undefined stoichiometry.</span>
0014 <span class="comment">%Type &quot;help getKEGGModelForOrganism&quot; to see what the different parameters are for.</span>
0015 <span class="comment">%This process takes from several hours to a couple of days, depending on</span>
0016 <span class="comment">%your hardware. The function is intended to be run on a cluster, so you can</span>
0017 <span class="comment">%speed up the process by running it on several computers simultaneously.</span>
0018 model=<a href="../external/kegg/getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,minScoreRatioKO,maxPhylDist,nSequences)">getKEGGModelForOrganism</a>(<span class="string">'sce'</span>,<span class="string">'sce.fa'</span>,<span class="string">'c:\input\eukaryota'</span>,<span class="string">'c:\output'</span>,false,false,false,10^-30,0.8,0.3,-1);
0019 
0020 <span class="comment">%As you can see the resulting model contains (around) 1081 reactions,</span>
0021 <span class="comment">%1182 metabolites and 809 genes. Small variations is possible since it is</span>
0022 <span class="comment">%an heuristic algorithm.</span>
0023 model
0024 
0025 
0026 <span class="comment">%A first control is that your model shouldn't be able to produce any metabolites</span>
0027 <span class="comment">%without uptake of some metabolites. This commonly happens when metabolites</span>
0028 <span class="comment">%have a different meaning in different reactions. The best way to find such</span>
0029 <span class="comment">%reactions is to run makeSomething and analyze the resulting solution for</span>
0030 <span class="comment">%bad reactions. An automated approach is to use removeBadRxns, which tries</span>
0031 <span class="comment">%to do the same thing in an automated manner. Please type &quot;help removeBadRxns&quot;</span>
0032 <span class="comment">%for details.</span>
0033 [newModel removedRxns]=<a href="../core/removeBadRxns.html" class="code" title="function [newModel removedRxns]=removeBadRxns(model,rxnRules,ignoreMets,isNames,balanceElements,refModel,ignoreIntBounds,printReport)">removeBadRxns</a>(model);
0034 
0035 <span class="comment">%You will see an error about that H+ can be made even if no reactions were</span>
0036 <span class="comment">%unbalanced. Protons are particularly problematic since it's rather</span>
0037 <span class="comment">%arbitary at which pH the formulas are written for. For the purpose of this</span>
0038 <span class="comment">%analysis we can ignore protons and try to fix it later.</span>
0039 [newModel removedRxns]=<a href="../core/removeBadRxns.html" class="code" title="function [newModel removedRxns]=removeBadRxns(model,rxnRules,ignoreMets,isNames,balanceElements,refModel,ignoreIntBounds,printReport)">removeBadRxns</a>(model,1,{<span class="string">'H+'</span>},true);
0040 
0041 <span class="comment">%Only one reaction was removed because it enabled the model to produce</span>
0042 <span class="comment">%something from nothing. Since there were so few, it might be worthwhile to look</span>
0043 <span class="comment">%into this in more detail.</span>
0044 removedRxns
0045 
0046 <span class="comment">%If you look it up in KEGG you will find that it is a general polymer reaction.</span>
0047 <span class="comment">%You might want to look at the flux distributions more in detail to try to</span>
0048 <span class="comment">%find out if there is any better alternative to delete. Use makeSomething</span>
0049 <span class="comment">%to do this</span>
0050 [fluxes metabolite]=<a href="../core/makeSomething.html" class="code" title="function [solution metabolite]=makeSomething(model,ignoreMets,isNames,minNrFluxes,allowExcretion,params,ignoreIntBounds)">makeSomething</a>(model,{<span class="string">'H+'</span>},true);
0051 model.metNames(metabolite)
0052 
0053 <span class="comment">%The model could produce H2O using the following reactions</span>
0054 <a href="../core/printFluxes.html" class="code" title="function printFluxes(model, fluxes, onlyExchange, cutOffFlux, outputFile,outputString,metaboliteList)">printFluxes</a>(model, fluxes, false, [], [],<span class="string">'%rxnID (%rxnName):\n\t%eqn: %flux\n'</span>)
0055 
0056 <span class="comment">%That resulted in quite a lot of fluxes to look through. Maybe it's easier</span>
0057 <span class="comment">%if we exclude the ones that were elementally balanced.</span>
0058 balanceStructure=<a href="../core/getElementalBalance.html" class="code" title="function balanceStructure=getElementalBalance(model,rxns,printUnbalanced,printUnparsable)">getElementalBalance</a>(model);
0059 
0060 <span class="comment">%We know that hydrogen balancing is a bit tricky, so let's only look at the</span>
0061 <span class="comment">%ones unbalanced for oxygen (since water was produced)</span>
0062 goodOnes=balanceStructure.leftComp(:,6)==balanceStructure.rightComp(:,6);
0063 <a href="../core/printFluxes.html" class="code" title="function printFluxes(model, fluxes, onlyExchange, cutOffFlux, outputFile,outputString,metaboliteList)">printFluxes</a>(removeRxns(model,goodOnes), fluxes(~goodOnes), false, [], [],<span class="string">'%rxnID (%rxnName):\n\t%eqn: %flux\n'</span>)
0064 
0065 <span class="comment">%That didn't really help, the only unbalanced reaction was the one that was</span>
0066 <span class="comment">%identified before. Let's print all fluxes involving amylose or starch</span>
0067 <span class="comment">%instead.</span>
0068 <a href="../core/printFluxes.html" class="code" title="function printFluxes(model, fluxes, onlyExchange, cutOffFlux, outputFile,outputString,metaboliteList)">printFluxes</a>(model, fluxes, false, [], [],<span class="string">'%rxnID (%rxnName):\n\t%eqn: %flux\n'</span>,{<span class="string">'Amylose'</span>;<span class="string">'Starch'</span>});
0069 
0070 <span class="comment">%This shows us two general and contradicting reactions. The first one says</span>
0071 <span class="comment">%that amylose and starch are interconvertible, the second one that you can</span>
0072 <span class="comment">%cleave of a glucose unit from starch to form amylose. This type of general</span>
0073 <span class="comment">%reactions are problematic and should be fixed manually. We therefore</span>
0074 <span class="comment">%choose to trust removeBadRxns and delete R02110.</span>
0075 model=removeRxns(model,<span class="string">'R02110'</span>);
0076 
0077 <span class="comment">%The model can no longer make something from nothing. Can it consume</span>
0078 <span class="comment">%something without any output?</span>
0079 [solution metabolite]=<a href="../core/consumeSomething.html" class="code" title="function [solution metabolite]=consumeSomething(model,ignoreMets,isNames,minNrFluxes,params,ignoreIntBounds)">consumeSomething</a>(model,{<span class="string">'H+'</span>},true);
0080 model.metNames(metabolite)
0081 
0082 <span class="comment">%Nope, so that was good. Let's add some uptakes and see what it can</span>
0083 <span class="comment">%produce.</span>
0084 [I J]=ismember({<span class="string">'D-Glucose'</span>;<span class="string">'H2O'</span>;<span class="string">'Orthophosphate'</span>;<span class="string">'Oxygen'</span>;<span class="string">'NH3'</span>;<span class="string">'Sulfate'</span>},model.metNames);
0085 [model addedRxns]=<a href="../core/addExchangeRxns.html" class="code" title="function [model addedRxns]=addExchangeRxns(model,reactionType,mets)">addExchangeRxns</a>(model,<span class="string">'in'</span>,J);
0086 
0087 <span class="comment">%Check which metabolites can be produced given these uptakes. The</span>
0088 <span class="comment">%canProduce function allows for output of all metabolites. This won't</span>
0089 <span class="comment">%happen in the real cell, but it is very useful for functionality testing</span>
0090 <span class="comment">%of the model. Once it is functional you can add excretion reactions based</span>
0091 <span class="comment">%on evidence.</span>
0092 I=<a href="../core/canProduce.html" class="code" title="function produced=canProduce(model,mets)">canProduce</a>(model);
0093 
0094 <span class="comment">%You can see that 31% of the metabolites could be synthesized. It is not</span>
0095 <span class="comment">%directly clear whether this is a high or low number, many metabolites</span>
0096 <span class="comment">%should not be possible to synthesize from those simple precursors.</span>
0097 sum(I)/numel(model.mets)
0098 
0099 <span class="comment">%We can try to fill gaps using the full KEGG model to see if that gives a</span>
0100 <span class="comment">%significantly higher number</span>
0101 keggModel=<a href="../external/kegg/getModelFromKEGG.html" class="code" title="function [model KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>([],false,false,false);
0102 
0103 <span class="comment">%The KEGG model is associated to some 800000 genes. They won't be used for</span>
0104 <span class="comment">%this, so we remove them to make this a little faster</span>
0105 keggModel=rmfield(keggModel,<span class="string">'genes'</span>);
0106 keggModel=rmfield(keggModel,<span class="string">'rxnGeneMat'</span>);
0107 
0108 <span class="comment">%We've already seen that there are some unbalanced reactions in KEGG. Let's</span>
0109 <span class="comment">%only use the balanced ones for the gap filling</span>
0110 balanceStructure=<a href="../core/getElementalBalance.html" class="code" title="function balanceStructure=getElementalBalance(model,rxns,printUnbalanced,printUnparsable)">getElementalBalance</a>(keggModel);
0111 keggModel=removeRxns(keggModel,balanceStructure.balanceStatus~=1,true,true);
0112 
0113 <span class="comment">%fillGaps with these settings will try to include reactions in order to</span>
0114 <span class="comment">%have flux through all reactions in the model. There are other settings as well.</span>
0115 <span class="comment">%The first flag says that production of all metabolites should be allowed.</span>
0116 params.relGap=0.6; <span class="comment">%Lower number for a more exhaustive search</span>
0117 params.printReport=true;
0118 [newConnected cannotConnect addedRxns newModel exitFlag]=<a href="../core/fillGaps.html" class="code" title="function [newConnected cannotConnect addedRxns newModel exitFlag]=fillGaps(model,models,allowNetProduction,useModelConstraints,supressWarnings,rxnScores,params)">fillGaps</a>(model,keggModel,true,false,false,[],params);
0119 
0120 <span class="comment">%We see that we could connect 48 reactions (newConnected) by including 54</span>
0121 <span class="comment">%reactions from the KEGG model (addedRxns). Those should of course be</span>
0122 <span class="comment">%checked manually to see that they exist in yeast, but let's say that we</span>
0123 <span class="comment">%have done so now.</span>
0124 
0125 <span class="comment">%You can now continue to improve the connectivity of the model by</span>
0126 <span class="comment">%identifying metabolites that should be connected. A convenient way to get</span>
0127 <span class="comment">%an overview of how connected the model is, and at the same time getting a</span>
0128 <span class="comment">%lot of useful data, is to use gapReport. Note that it can take several</span>
0129 <span class="comment">%hours to run, depending on the number of gaps in the model.</span>
0130 [noFluxRxns noFluxRxnsRelaxed subGraphs notProducedMets minToConnect<span class="keyword">...</span>
0131     neededForProductionMat]=<a href="../core/gapReport.html" class="code" title="function [noFluxRxns, noFluxRxnsRelaxed, subGraphs, notProducedMets, minToConnect,neededForProductionMat, canProduceWithoutInput, canConsumeWithoutOutput,connectedFromTemplates, addedFromTemplates]=gapReport(model, templateModels)">gapReport</a>(newModel);
0132 
0133 <span class="comment">%You will see that 361/1140 reactions cannot carry flux. Remember that we</span>
0134 <span class="comment">%allow for output of all metabolites, that's why it calculates 361 in both</span>
0135 <span class="comment">%cases. 557/1195 metabolites cannot be synthesized from the precursors we have</span>
0136 <span class="comment">%supplied. There are 7 subnetworks in the model, of which 1181/1195</span>
0137 <span class="comment">%metabolites belong to the first one.</span>
0138 <span class="comment">%</span>
0139 <span class="comment">%It will also print something similar to:</span>
0140 <span class="comment">%</span>
0141 <span class="comment">%To enable net production of all metabolites, a total of 266 metabolites</span>
0142 <span class="comment">%must be connected</span>
0143 <span class="comment">%Top 10 metabolites to connect:</span>
0144 <span class="comment">%     1. Acyl-[acyl-carrier protein][s] (connects 65 metabolites)</span>
0145 <span class="comment">%     2. 1-Radyl-2-acyl-sn-glycero-3-phosphocholine[s] (connects 16 metabolites)</span>
0146 <span class="comment">%     3. G00146[s] (connects 13 metabolites)</span>
0147 <span class="comment">%     4. NAD+[s] (connects 12 metabolites)</span>
0148 <span class="comment">%     5. Tetrahydrofolate[s] (connects 11 metabolites)</span>
0149 <span class="comment">%     6. Methylselenic acid[s] (connects 10 metabolites)</span>
0150 <span class="comment">%     7. G00003[s] (connects 10 metabolites)</span>
0151 <span class="comment">%     8. G00009[s] (connects 8 metabolites)</span>
0152 <span class="comment">%     9. FAD[s] (connects 6 metabolites)</span>
0153 <span class="comment">%     10. Thiamin monophosphate[s] (connects 6 metabolites)</span>
0154 
0155 <span class="comment">%This is a very useful way of directing the gap-filling tasks to where they</span>
0156 <span class="comment">%are of the greatest use. In this case it says that in order to have net</span>
0157 <span class="comment">%production of all metabolites in the model from the simple inputs that we have</span>
0158 <span class="comment">%defined (557 metabolites can currently not have net production) we need to</span>
0159 <span class="comment">%connect 226 unconnected metabolites in. However, by connecting only the</span>
0160 <span class="comment">%top 10 in the list 157/557 (28%) of them will now be connected.</span>
0161 <span class="comment">%If we look at the list we see that the top ranking one is</span>
0162 <span class="comment">%Acyl-[acyl-carrier protein]. This is because acyl-carrier protein acts as</span>
0163 <span class="comment">%a carrier for many fatty acids, but it is not synthesized in the model</span>
0164 <span class="comment">%from its amino acid constituents. Therefore there cannot be net production</span>
0165 <span class="comment">%of any of the metabolites involving it. One solution would be to add a</span>
0166 <span class="comment">%reaction representing the synthesis of it (based on its amino acid</span>
0167 <span class="comment">%composition). Another solution would be to add an uptake reaction for it.</span>
0168 <span class="comment">%Metabolites 4,5,8, and 9 are all co-factors or involved in co-factor</span>
0169 <span class="comment">%metabolism. It turns out that yeast cannot grow on only the substrates we</span>
0170 <span class="comment">%have defined, but that it requires some other precursors for co-factor</span>
0171 <span class="comment">%synthesis as well.</span>
0172 
0173 <span class="comment">%Add uptake reactions for the minimal media constituents needed for yeast</span>
0174 <span class="comment">%to grow.</span>
0175 [I J]=ismember({<span class="string">'4-Aminobenzoate'</span>;<span class="string">'Riboflavin'</span>;<span class="string">'Thiamine'</span>;<span class="string">'Biotin'</span>;<span class="string">'Folate'</span>;<span class="string">'Nicotinate'</span>;<span class="string">'Zymosterol'</span>;<span class="string">'Choline'</span>},newModel.metNames);
0176 [newModel addedRxns]=<a href="../core/addExchangeRxns.html" class="code" title="function [model addedRxns]=addExchangeRxns(model,reactionType,mets)">addExchangeRxns</a>(newModel,<span class="string">'in'</span>,J);
0177 
0178 <span class="comment">%Rerun gapReport and use the output for targeting the gap-filling efforts.</span>
0179 <span class="comment">%Note that only some info is printed; most of it is available in the output</span>
0180 <span class="comment">%structures.</span>
0181 <span class="comment">%Work like this in an iterative manner until the model is of sufficient</span>
0182 <span class="comment">%quality.</span></pre></div>
<hr><address>Generated on Fri 14-Oct-2016 12:02:10 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>